### PR TITLE
Spring boot upgrade and JSON mapping issue

### PR DESCRIPTION
**Pull request by vapadwal**
Fix for issue 
https://github.com/devonfw/devon4j/issues/12
https://github.com/devonfw/devon4j/issues/106

Fix does not include IdRef

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -27,7 +27,7 @@
     <changelist>-SNAPSHOT</changelist>
     <devon4j.version>${revision}${changelist}</devon4j.version>
     <devon4j.flatten.mode>oss</devon4j.flatten.mode>
-    <spring.boot.version>2.1.2.RELEASE</spring.boot.version>
+    <spring.boot.version>2.1.6.RELEASE</spring.boot.version>
```
Comment (hohwille) [inline pom.xml]: OK. so you also fixed the version update in the same PR.
Fine.

File: pom.xml [hunk 2]
```diff
@@ -341,7 +341,7 @@
         <plugin>
           <groupId>org.springframework.boot</groupId>
           <artifactId>spring-boot-maven-plugin</artifactId>
-          <version>2.0.4.RELEASE</version>
+          <version>2.1.6.RELEASE</version>
```
Comment (hohwille) [inline pom.xml]: Nice catch. But could you be so nice to use `${spring.boot.version}` so we get this aligned also in the future.

File: templates/server/src/main/resources/archetype-resources/pom.xml [hunk 3]
```diff
@@ -287,7 +287,7 @@
         <plugin>
           <groupId>org.springframework.boot</groupId>
           <artifactId>spring-boot-maven-plugin</artifactId>
-          <version>2.0.4.RELEASE</version>
+          <version>2.1.6.RELEASE</version>
```
Comment (hohwille) [inline templates/server/src/main/resources/archetype-resources/pom.xml]: fine but same as above: should be `${spring.boot.version}`

File: templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/common/impl/config/ApplicationObjectMapperFactory.java [hunk 4]
```diff
@@ -32,6 +34,9 @@ public ApplicationObjectMapperFactory() {
     // register spring-data Pageable
     module.addSerializer(Pageable.class, new PageableJsonSerializer());
     module.addDeserializer(Pageable.class, new PageableJsonDeserializer());
+    // register spring-data Page
+    module.addAbstractTypeMapping(Page.class, JsonPage.class);
+    module.setMixInAnnotation(Page.class, JsonPage.class);
```
Comment (hohwille) [inline templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/common/impl/config/ApplicationObjectMapperFactory.java]: fine, fixes our problem.
As I already stated we should IMHO provide the fundamental config in `ObjectMapperFactory` in `devon4j` module in a protected method. We have done things wrong here several times and bringing out bugfixes to the archetype does not help the projects that have already started from the archetype.
IMHO all what is configured here could be done in a new projected method called e.g. `ObjectMapperFactory.initMapping()`.

File: pom.xml (hunk 1 above)
Comment (hohwille) [reply pom.xml]: I now see you also explained that in the top-level comment of this PR. Excellent.

File: pom.xml (hunk 2 above)
Comment (vapadwal) [reply pom.xml]: > @vapadwal 👍 thanks. You solved the issue correct. We could already merge it this way. However, I added some comments for further improvements. As they should be trivial to implement, I would love to ask you to improve this PR.
> 
> Also we should add more tests (e.g. deserialization of Page, Pageable, Instant, Year, CsrfToken, etc.). I am fine if you can not include them on time into this PR but could you then be so kind to file an issue to add more tests for ObjectMapperFactory.

Issue added
https://github.com/devonfw/devon4j/issues/112

File: templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/common/impl/config/ApplicationObjectMapperFactory.java (hunk 4 above)
Comment (vapadwal) [reply templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/common/impl/config/ApplicationObjectMapperFactory.java]: Done

File: templates/server/src/main/resources/archetype-resources/pom.xml (hunk 3 above)
Comment (vapadwal) [reply templates/server/src/main/resources/archetype-resources/pom.xml]: Done

File: modules/json/src/main/java/com/devonfw/module/json/common/base/ObjectMapperFactory.java [hunk 5]
```diff
@@ -141,4 +147,17 @@ public ObjectMapper createInstance() {
 
     return mapper;
   }
+
+  /**
+   * @return an instance of {@link SimpleModule} for registering configurations
+   */
+  protected SimpleModule initMapping() {
```
Comment (hohwille) [inline modules/json/src/main/java/com/devonfw/module/json/common/base/ObjectMapperFactory.java]: great. Just call it from the constructor of this ObjectMapperFactory.

File: templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/common/impl/config/ApplicationObjectMapperFactory.java [hunk 6]
```diff
@@ -27,11 +25,8 @@ public ApplicationObjectMapperFactory() {
 
     super();
     // see https://github.com/devonfw-wiki/devon4j/wiki/guide-json#json-and-inheritance
-    SimpleModule module = getExtensionModule();
+    SimpleModule module = super.initMapping();
```
Comment (hohwille) [inline templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/common/impl/config/ApplicationObjectMapperFactory.java]: when `initMapping()` is called from constructor, we can keep the original `getExtensionModule()` line.

File: modules/json/src/main/java/com/devonfw/module/json/common/base/ObjectMapperFactory.java (hunk 5 above)
Comment (vapadwal) [reply modules/json/src/main/java/com/devonfw/module/json/common/base/ObjectMapperFactory.java]: Done

File: templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/common/impl/config/ApplicationObjectMapperFactory.java (hunk 6 above)
Comment (vapadwal) [reply templates/server/src/main/resources/archetype-resources/core/src/main/java/__packageInPathFormat__/general/common/impl/config/ApplicationObjectMapperFactory.java]: Done

**Pull request merged**